---
title: "Splicing influence estimation by Long Reads: SaiLoR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SaiLoR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, warning=FALSE, message=FALSE}
library(SaiLoR)
```

# Running SAILOR 

# Required input files: 
  * bam files [minimap2](https://github.com/lh3/minimap2) 
  * reference annotation 
  * short read sequencing SJ.out files from [STAR](https://github.com/alexdobin/STAR)

# Building junction reference. 

SaiLoR requires a reference junction set to identify couplings. It only analyzes junctions that occur outside the influence of 5' selection (alternative transcription start sites) or 3' selection (alternative polyadenylation). Sailor can build the reference using both a reference annotation and short read sequence data junctions from STAR.

```{r, warning=FALSE, message=FALSE}
# load reference annotation 
annot_path <- system.file("inst/exdata", "dm6.annot.gtf.gz", package="SaiLoR")
ref_annot <- rtracklayer::import.gff(annot_path)
# load junctions from STAR 
junction_path <- system.file("inst/exdata","short_read_junctions.SJ.out.tab", package = 'SaiLoR')
# create txf file of reference junctions 
referenece_junctions <- create_reference_junctions( junction_path, min.jcounts = 2 , ref_annot, type="short")
referenece_junctions
```

![SaiLoR regions of interest](inst/exdata/exonAnalysis.jpg) 

# Read to junctions assignments 

## Full-length read filtering. 

SaiLoR generates a database with read-to-feature assignments. This process involves three steps occurring within `read_to_junctions()`. The steps include: 1) Identifying only 5'-3' full length read isoforms. This filtering is done with `get_sailor_full_lengths()`, considering only reads spanning an annotated TSS and 3' end. 2) Correcting the junctions found in long reads using reference annotation and short read data in `read_refjunction_correction()`. 3) Last, `make_junction_database()` takes each read and assigns the features it contains: TSS, TES, and the junctions it contains.  

```{r,warning=FALSE, message=FALSE}
bamPath <- system.file("inst/exdata","testBam.bam", package = 'SaiLoR')
exonlinks.counts<- read_to_junctions(bamPath, referenece_junctions, annot_path)
```

# Calculate exon-junction couplings to 5'/3'

SaiLoR calculates exon couplings by generating contingency matrices from exon junctions to promoters with `compute_exon_tss_couplings()` and exon junctions to 3' ends with `compute_exon_3end_couplings()`. It uses multinomial testing with the chi-squared statistic to test the significance of the couplings. To account for possible biases due to low counts, the p-value is simulated using Monte Carlo methods.

```{r,warning=FALSE, message=FALSE}
couplings <- calculate_exon_couplings(exonlinks.counts,  referenece_junctions )
```









